### PR TITLE
Updating the HWIntrinsics to support dropping unnecessary casts

### LINKED ISSUE #36254: Vector64 and Vector128.Create doesn't generate movi for long immediates
`Vector64.Create` and `Vector128.Create` don't generate `movi` as the other types do for immediate inputs. This appears to be because the logic is using `IsCnsIntOrI` whioch will miss `GT_CNS_LNG`: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lowerarmarch.cpp#L933-L957

CC. @kunalspathak, @CarolEidt, @echesakovMSFT

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/36254 and cleans up the more general issue of unnecessary `GT_CAST` for a couple other intrinsics.

### CODE REVIEW

**Comment by tannergooding** [comment]
x64 Frameworks Diff:
```
Found 267 files with textual diffs.

Summary of Code Size diffs:
(Lower is better)

Total bytes of diff: -204 (-0.000% of base)
    diff is an improvement.

Top file improvements (bytes):
        -204 : System.Private.CoreLib.dasm (-0.004% of base)

1 total files with Code Size differences (1 improved, 0 regressed), 265 unchanged.

Top method regressions (bytes):
         252 (9.848% of base) : System.Private.CoreLib.dasm - Sse2:Insert(Vector128`1,short,ubyte):Vector128`1
         252 (9.848% of base) : System.Private.CoreLib.dasm - Sse2:Insert(Vector128`1,ushort,ubyte):Vector128`1

Top method improvements (bytes):
        -176 (-34.783% of base) : System.Private.CoreLib.dasm - Vector256:Create(byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte):Vector256`1
        -148 (-31.092% of base) : System.Private.CoreLib.dasm - Vector256:Create(ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte):Vector256`1
         -86 (-37.885% of base) : System.Private.CoreLib.dasm - Vector128:Create(byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte):Vector128`1
         -78 (-35.455% of base) : System.Private.CoreLib.dasm - Vector256:Create(short,short,short,short,short,short,short,short,short,short,short,short,short,short,short,short):Vector256`1
         -73 (-34.272% of base) : System.Private.CoreLib.dasm - Vector128:Create(ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte):Vector128`1
         -66 (-32.039% of base) : System.Private.CoreLib.dasm - Vector256:Create(ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort):Vector256`1
         -36 (-35.294% of base) : System.Private.CoreLib.dasm - Vector128:Create(short,short,short,short,short,short,short,short):Vector128`1
         -31 (-32.292% of base) : System.Private.CoreLib.dasm - Vector128:Create(ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort):Vector128`1
          -4 (-0.142% of base) : System.Private.CoreLib.dasm - Sse41:Insert(Vector128`1,byte,ubyte):Vector128`1
          -4 (-0.142% of base) : System.Private.CoreLib.dasm - Sse41:Insert(Vector128`1,ubyte,ubyte):Vector128`1
          -3 (-27.273% of base) : System.Private.CoreLib.dasm - Sse42:Crc32(int,ubyte):int
          -3 (-25.000% of base) : System.Private.CoreLib.dasm - Sse42:Crc32(int,ushort):int

Top method regressions (percentages):
         252 (9.848% of base) : System.Private.CoreLib.dasm - Sse2:Insert(Vector128`1,short,ubyte):Vector128`1
         252 (9.848% of base) : System.Private.CoreLib.dasm - Sse2:Insert(Vector128`1,ushort,ubyte):Vector128`1

Top method improvements (percentages):
         -86 (-37.885% of base) : System.Private.CoreLib.dasm - Vector128:Create(byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte):Vector128`1
         -78 (-35.455% of base) : System.Private.CoreLib.dasm - Vector256:Create(short,short,short,short,short,short,short,short,short,short,short,short,short,short,short,short):Vector256`1
         -36 (-35.294% of base) : System.Private.CoreLib.dasm - Vector128:Create(short,short,short,short,short,short,short,short):Vector128`1
        -176 (-34.783% of base) : System.Private.CoreLib.dasm - Vector256:Create(byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte,byte):Vector256`1
         -73 (-34.272% of base) : System.Private.CoreLib.dasm - Vector128:Create(ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte):Vector128`1
         -31 (-32.292% of base) : System.Private.CoreLib.dasm - Vector128:Create(ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort):Vector128`1
         -66 (-32.039% of base) : System.Private.CoreLib.dasm - Vector256:Create(ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort,ushort):Vector256`1
        -148 (-31.092% of base) : System.Private.CoreLib.dasm - Vector256:Create(ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte,ubyte):Vector256`1
          -3 (-27.273% of base) : System.Private.CoreLib.dasm - Sse42:Crc32(int,ubyte):int
          -3 (-25.000% of base) : System.Private.CoreLib.dasm - Sse42:Crc32(int,ushort):int
          -4 (-0.142% of base) : System.Private.CoreLib.dasm - Sse41:Insert(Vector128`1,byte,ubyte):Vector128`1
          -4 (-0.142% of base) : System.Private.CoreLib.dasm - Sse41:Insert(Vector128`1,ubyte,ubyte):Vector128`1

14 total methods with Code Size differences (12 improved, 2 regressed), 244908 unchanged.
```

The biggeset win is:
```diff
        movzx    rax, dl
        vmovd    xmm0, eax
-       movzx    rax, r8b
-       vpinsrb  xmm0, xmm0, xrax, 1
-       movzx    rax, r9b
-       vpinsrb  xmm0, xmm0, xrax, 2
-       mov      eax, dword ptr [rsp+28H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 3
-       mov      eax, dword ptr [rsp+30H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 4
-       mov      eax, dword ptr [rsp+38H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 5
-       mov      eax, dword ptr [rsp+40H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 6
-       mov      eax, dword ptr [rsp+48H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 7
-       mov      eax, dword ptr [rsp+50H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 8
-       mov      eax, dword ptr [rsp+58H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 9
-       mov      eax, dword ptr [rsp+60H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 10
-       mov      eax, dword ptr [rsp+68H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 11
-       mov      eax, dword ptr [rsp+70H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 12
-       mov      eax, dword ptr [rsp+78H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 13
-       mov      eax, dword ptr [rsp+80H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 14
-       mov      eax, dword ptr [rsp+88H]
-       movzx    rax, al
-       vpinsrb  xmm0, xmm0, xrax, 15
+       vpinsrb  xmm0, xmm0, xr8, 1
+       vpinsrb  xmm0, xmm0, xr9, 2
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+28H], 3
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+30H], 4
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+38H], 5
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+40H], 6
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+48H], 7
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+50H], 8
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+58H], 9
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+60H], 10
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+68H], 11
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+70H], 12
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+78H], 13
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+80H], 14
+       vpinsrb  xmm0, xmm0, xmmword ptr [rsp+88H], 15
        vmovupd  xmmword ptr [rcx], xmm0
        mov      rax, rcx
-						;; bbWeight=1    PerfScore 51.25
+						;; bbWeight=1    PerfScore 21.50
```

Other diffs are similar. The `Sse2:Insert` regression is for the indirect invocation callback where we have the giant jump table to handle non-immediate inputs. Previously the sign extension was moving the input operand from `r8` to `rax` which resulted in saving a byte per switch case.

The diff for CRC32 is:
```asm
-       movzx    rax, dl
-       crc32    ecx, al
        mov      eax, ecx
+       crc32    eax, dl
-						;; bbWeight=1    PerfScore 2.50
+						;; bbWeight=1    PerfScore 2.25
```

**Comment by tannergooding** [comment]
ARM64 Framework Diff:
```
Found 270 files with textual diffs.

Summary of Code Size diffs:
(Lower is better)

Total bytes of diff: -8 (-0.000% of base)
    diff is an improvement.

Top file improvements (bytes):
          -8 : System.Collections.dasm (-0.001% of base)

1 total files with Code Size differences (1 improved, 0 regressed), 265 unchanged.

Top method improvements (bytes):
          -8 (-4.444% of base) : System.Collections.dasm - BitArray:.cctor()

Top method improvements (percentages):
          -8 (-4.444% of base) : System.Collections.dasm - BitArray:.cctor()

1 total methods with Code Size differences (1 improved, 0 regressed), 244921 unchanged.
```

We see a smaller diff overall due to instructions having the same size encoding. Most of the diffs are cases like:
```diff
-            ldr     q16, [@RWD00]
+            mvni    v16.4s, #0x00
             mov     v0.16b, v16.16b
-						;; bbWeight=1    PerfScore 2.50
+						;; bbWeight=1    PerfScore 1.00
```

**Comment by tannergooding** [comment]
X64 Tests Diff:
```
Found 3414 files with textual diffs.

Summary of Code Size diffs:
(Lower is better)

Total bytes of diff: -77 (-0.000% of base)
    diff is an improvement.

Top file improvements (bytes):
         -22 : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm (-0.001% of base)
         -22 : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm (-0.003% of base)
         -12 : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm (-0.014% of base)
         -10 : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_r\GitHub_21666_r.dasm (-0.239% of base)
          -8 : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_ro\GitHub_21666_ro.dasm (-0.362% of base)
          -3 : JIT\HardwareIntrinsics\X86\Regression\GitHub_23438\GitHub_23438_r\GitHub_23438_r.dasm (-0.083% of base)

6 total files with Code Size differences (6 improved, 0 regressed), 3320 unchanged.

Top method improvements (bytes):
          -6 (-3.333% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertInt161:RunStructLclFldScenario():this
          -6 (-3.333% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertInt16129:RunStructLclFldScenario():this
          -6 (-3.315% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertSByte1:RunStructLclFldScenario():this
          -6 (-3.315% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertSByte129:RunStructLclFldScenario():this
          -5 (-2.809% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertUInt161:RunStructLclFldScenario():this
          -5 (-2.809% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertUInt16129:RunStructLclFldScenario():this
          -5 (-2.793% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertByte1:RunStructLclFldScenario():this
          -5 (-2.793% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertByte129:RunStructLclFldScenario():this
          -5 (-0.762% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_r\GitHub_21666_r.dasm - GitHub_21666:TestByteContainment():bool
          -5 (-0.755% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_r\GitHub_21666_r.dasm - GitHub_21666:TestUInt16Containment():bool
          -4 (-1.087% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_ro\GitHub_21666_ro.dasm - GitHub_21666:TestByteContainment():bool
          -4 (-1.075% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_ro\GitHub_21666_ro.dasm - GitHub_21666:TestUInt16Containment():bool
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertInt641:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertUInt641:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertInt64129:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertUInt64129:RunBasicScenario():this
          -3 (-1.744% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_23438\GitHub_23438_r\GitHub_23438_r.dasm - Program:TestSse41X64Insert_UInt64():bool

Top method improvements (percentages):
          -6 (-3.333% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertInt161:RunStructLclFldScenario():this
          -6 (-3.333% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertInt16129:RunStructLclFldScenario():this
          -6 (-3.315% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertSByte1:RunStructLclFldScenario():this
          -6 (-3.315% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertSByte129:RunStructLclFldScenario():this
          -5 (-2.809% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertUInt161:RunStructLclFldScenario():this
          -5 (-2.809% of base) : JIT\HardwareIntrinsics\X86\Sse2\Sse2_ro\Sse2_ro.dasm - InsertScalarTest__InsertUInt16129:RunStructLclFldScenario():this
          -5 (-2.793% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertByte1:RunStructLclFldScenario():this
          -5 (-2.793% of base) : JIT\HardwareIntrinsics\X86\Sse41\Sse41_ro\Sse41_ro.dasm - InsertScalarTest__InsertByte129:RunStructLclFldScenario():this
          -3 (-1.744% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_23438\GitHub_23438_r\GitHub_23438_r.dasm - Program:TestSse41X64Insert_UInt64():bool
          -4 (-1.087% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_ro\GitHub_21666_ro.dasm - GitHub_21666:TestByteContainment():bool
          -4 (-1.075% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_ro\GitHub_21666_ro.dasm - GitHub_21666:TestUInt16Containment():bool
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertInt641:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertUInt641:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertInt64129:RunBasicScenario():this
          -3 (-0.893% of base) : JIT\HardwareIntrinsics\X86\Sse41.X64\Sse41.X64_r\Sse41.X64_r.dasm - InsertScalarTest__InsertUInt64129:RunBasicScenario():this
          -5 (-0.762% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_r\GitHub_21666_r.dasm - GitHub_21666:TestByteContainment():bool
          -5 (-0.755% of base) : JIT\HardwareIntrinsics\X86\Regression\GitHub_21666\GitHub_21666_r\GitHub_21666_r.dasm - GitHub_21666:TestUInt16Containment():bool

17 total methods with Code Size differences (17 improved, 0 regressed), 350671 unchanged.

82 files had text diffs but no metric diffs.
JIT\HardwareIntrinsics\X86\Avx2\GatherMaskVector128_ro\GatherMaskVector128_ro.dasm had 4 diffs
JIT\HardwareIntrinsics\X86\Avx2\GatherMaskVector256_ro\GatherMaskVector256_ro.dasm had 4 diffs
Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest\TestLibrary.dasm had 2 diffs
Interop\DllImportAttribute\ExactSpelling\ExactSpellingTest\TestLibrary.dasm had 2 diffs
Interop\ICustomMarshaler\ConflictingNames\CustomMarshaler\TestLibrary.dasm had 2 diffs
Interop\ICustomMarshaler\ConflictingNames\CustomMarshaler2\TestLibrary.dasm had 2 diffs
Interop\ICustomMarshaler\ConflictingNames\MultipleALCs\TestLibrary.dasm had 2 diffs
Interop\IJW\CopyConstructorMarshaler\CopyConstructorMarshaler\TestLibrary.dasm had 2 diffs
Interop\IJW\FixupCallsHostWhenLoaded\FixupCallsHostWhenLoaded\TestLibrary.dasm had 2 diffs
Interop\IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.dasm had 2 diffs
Interop\IJW\ManagedCallingNative\ManagedCallingNative\TestLibrary.dasm had 2 diffs
Interop\IJW\NativeCallingManaged\NativeCallingManaged\TestLibrary.dasm had 2 diffs
Interop\PInvoke\CustomMarshalers\CustomMarshalersTest\CustomMarshalersTest.dasm had 2 diffs
Interop\StructMarshalling\PInvoke\MarshalStructAsLayoutSeq\TestLibrary.dasm had 2 diffs
Interop\COM\ComWrappers\WeakReference\WeakReferenceTest\TestLibrary.dasm had 2 diffs
Interop\COM\NETClients\Aggregation\NETClientAggregation\TestLibrary.dasm had 2 diffs
Interop\COM\NETClients\ConsumeNETServer\ConsumeNETServer\TestLibrary.dasm had 2 diffs
Interop\COM\NETClients\Primitives\NETClientPrimitives\TestLibrary.dasm had 2 diffs
Interop\COM\NETClients\Primitives\NETClientPrimitivesInALC\TestLibrary.dasm had 2 diffs
Interop\PInvoke\CriticalHandles\ReverseTest\ReverseTest\TestLibrary.dasm had 2 diffs
```

**Comment by tannergooding** [comment]
CC. @CarolEidt, @kunalspathak, @echesakovMSFT

**Comment by tannergooding** [comment]
Merging in dotnet/master so this doesn't get out of date.
It is ready for review.

File: src/coreclr/src/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -947,6 +947,61 @@ void Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                argList->Rest()->gtOp1 = castOp;
+            }
+            break;
+        }
+
+        case NI_SSE42_Crc32:
+        {
+            assert(HWIntrinsicInfo::lookupNumArgs(node) == 2);
+
+            GenTree* op2 = node->gtOp2;
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lowerxarch.cpp:983]: It's not a lot of code, but it's unfortunate to duplicate this. You might consider creating a method that takes the possible cast op, and then returns it or its child.

**Review by CarolEidt — APPROVED** [review]
One optional de-duplication suggestion, but LGTM.

File: src/coreclr/src/jit/lowerxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/src/jit/lowerxarch.cpp:983]: Are you thinking something like:
```csharp
node->gtOp2 = RemoveCastIfPresent(node->gtType, node->gtOp2);
```

and
```csharp
GenTree* RemoveCastIfPresent(var_types expectedType, GenTree* op)
{
    if (!op->OperIs(GT_CAST))
    {
        return op;
    }

    GenTree* castOp = op2->AsCast()->CastOp();

    if (genTypeSize(castOp->gtType) >= genTypeSize(expectedType))
    {
        BlockRange().Remove(op);
        return castOp;
    }

    return op;
}
```

?

Comment (tannergooding) [reply src/coreclr/src/jit/lowerxarch.cpp:983]: Maybe `TryRemoveCast` (same signature) would be a better name?

**Review by kunalspathak — APPROVED** [review]

File: src/coreclr/src/jit/lowerxarch.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/coreclr/src/jit/lowerxarch.cpp:983]: Yes, something like that was what I was thinking. As I say, though, it's not a lot of code so I don't want to block you on it.

Comment (tannergooding) [reply src/coreclr/src/jit/lowerxarch.cpp:983]: Sounds good! I've basically got it done locally (just doing some sanity testing) so I'll merge this and will get a follow up PR shortly after.

The downside on ARM64 is that the existing use-case isn't easy to share. It is conditionalized on the underlying operation being a constant and being within a specific range. But I imagine there may be other use-cases that can share in the future

**Review by echesakov — APPROVED** [review]

**Pull request merged**